### PR TITLE
C# SFTP Client

### PR DESCRIPTION
**Pull request by lalchanp**
As discussed here is the C# code for SFTP File Transfers.

Refer readme.md file for documentation and dependencies.

### CODE REVIEW

**Comment by motymichaely** [comment]
@paras-lalchandani Thanks for your contribution!  

Can you please make the following changes?
1. Restructure the folders to be under `dotnet/src` instead of `c#`
2. Rename `readme.md` to `README.md` and put it under `dotnet` folder.

**Comment by lalchanp** [comment]
> @paras-lalchandani Thanks for your contribution!
> 
> Can you please make the following changes?
> 
> 1. Restructure the folders to be under `dotnet/src` instead of `c#`
> 2. Rename `readme.md` to `README.md` and put it under `dotnet` folder.

@motymichaely Completed above suggestions

File: dotnet/src/SFTPService/SftpService.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+            using var client = new SftpClient(_config.Host, _config.Port == 0 ? 22 : _config.Port, _config.UserName, _config.Password);
+            try
+            {
+                client.Connect();
+                client.DeleteFile(remoteFilePath);
+                _logger.LogInformation($"File [{remoteFilePath}] deleted.");
+            }
+            catch (Exception exception)
+            {
+                _logger.LogError(exception, $"Failed in deleting file [{remoteFilePath}]");
```
Comment (motymichaely) [inline dotnet/src/SFTPService/SftpService.cs]: @paras-lalchandani Should be `Failed to delete file`

File: dotnet/README.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
+## Getting Started with SFTP using C#(.NET)
+
+#Project Description
+
+The project contains two solutions:
+- SFTPClient: This is a simple console application which depicts how we can use our SFTP service to perform differnet file operations.
+- SFPTService: This is a simple class library project that conatains the contains code to connect, upload , dowload and delete files.
```
Comment (motymichaely) [inline dotnet/README.md]: @paras-lalchandani Fix typos:

`This is a simple class library project that contains code to connect, upload ,download and delete files.

File: dotnet/README.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+			public class SftpConfig
+			{
+				public string Host { get; set; }
+				public int Port { get; set; }
+				public string UserName { get; set; }
+				public string Password { get; set; }
+			}
+	```
+
+3. The next step is to simply call our SFTP service.Any operation performed inside the SFTP service would have :
```
Comment (motymichaely) [inline dotnet/README.md]: @paras-lalchandani  Some typos: `The next step is to simply call our SFTP service. Any operation performed inside the SFTP service would have to:`

File: dotnet/README.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+FROM:
+
+using var client = new SftpClient(_config.Host, _config.Port == 0 ? 22 : _config.Port, _config.UserName, _config.Password);
+
+TO: 
+
+var privateKey = new PrivateKeyFile(@"C:\some\path\key.pem");
+using var client = new SftpClient(_config.Host, _config.UserName, new[] { privateKey });
+
+where (@"C:\some\path\key.pem") defines the path to your keys
```
Comment (motymichaely) [inline dotnet/README.md]: @paras-lalchandani Please replace `to your keys` to `to your public key`

File: dotnet/src/SFTPService/SftpService.cs [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        {
+            using var client = new SftpClient(_config.Host, _config.Port == 0 ? 22 : _config.Port, _config.UserName, _config.Password);
+            try
+            {
+                client.Connect();
+                return client.ListDirectory(remoteDirectory);
+            }
+            catch (Exception exception)
+            {
+                _logger.LogError(exception, $"Failed in listing files under [{remoteDirectory}]");
```
Comment (motymichaely) [inline dotnet/src/SFTPService/SftpService.cs]: @paras-lalchandani `Failed to list files under`

File: dotnet/src/SFTPService/SftpService.cs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            try
+            {
+                client.Connect();
+                using var s = File.OpenRead(localFilePath);
+                client.UploadFile(s, remoteFilePath);
+                _logger.LogInformation($"Finished uploading file [{localFilePath}] to [{remoteFilePath}]");
+            }
+            catch (Exception exception)
+            {
+                _logger.LogError(exception, $"Failed in uploading file [{localFilePath}] to [{remoteFilePath}]");
```
Comment (motymichaely) [inline dotnet/src/SFTPService/SftpService.cs]: @paras-lalchandani `Failed to upload file`

File: dotnet/src/SFTPService/SftpService.cs [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+            try
+            {
+                client.Connect();
+                using var s = File.Create(localFilePath);
+                client.DownloadFile(remoteFilePath, s);
+                _logger.LogInformation($"Finished downloading file [{localFilePath}] from [{remoteFilePath}]");
+            }
+            catch (Exception exception)
+            {
+                _logger.LogError(exception, $"Failed in downloading file [{localFilePath}] from [{remoteFilePath}]");
```
Comment (motymichaely) [inline dotnet/src/SFTPService/SftpService.cs]: @paras-lalchandani `Failed to download file`

**Review by motymichaely — CHANGES_REQUESTED** [review]
@paras-lalchandani Thanks for your recent changes!  I've added a few comments, mainly typos. Please change and push your changes for a final review.

File: dotnet/src/SFTPService/SftpService.cs (hunk 7 above)
Comment (lalchanp) [reply dotnet/src/SFTPService/SftpService.cs]: @motymichaely done

File: dotnet/src/SFTPService/SftpService.cs (hunk 6 above)
Comment (lalchanp) [reply dotnet/src/SFTPService/SftpService.cs]: @motymichaely done

File: dotnet/src/SFTPService/SftpService.cs (hunk 5 above)
Comment (lalchanp) [reply dotnet/src/SFTPService/SftpService.cs]: @motymichaely done

File: dotnet/README.md (hunk 4 above)
Comment (lalchanp) [reply dotnet/README.md]: @motymichaely done

File: dotnet/README.md (hunk 3 above)
Comment (lalchanp) [reply dotnet/README.md]: @motymichaely done

File: dotnet/README.md (hunk 2 above)
Comment (lalchanp) [reply dotnet/README.md]: @motymichaely done

File: dotnet/src/SFTPService/SftpService.cs (hunk 1 above)
Comment (lalchanp) [reply dotnet/src/SFTPService/SftpService.cs]: @motymichaely done

**Comment by lalchanp** [comment]
@motymichaely Completed the latest typo issue . Please reverify

**Comment by lalchanp** [comment]
Completed the suggested changes .

**Review by motymichaely — APPROVED** [review]

**Pull request merged**